utcp.c

```diff
@@ -635,63 +893,195 @@ ssize_t utcp_recv(struct utcp *utcp, const void *data, size_t len) {
 		break;
 	default:
 #ifdef UTCP_DEBUG
+		debug("Error: utcp_recv unexpected connection state %p %s\n", c, strstate[c->state]);
 		abort();
 #endif
 		break;
 	}
 
-	// 1b. Drop packets with a sequence number not in our receive window.
+	// 1b. Drop packets with an invalid ACK.
+	// ackno should not roll back, and it should also not be bigger than snd.last,
+	// but it might be bigger than snd.nxt since we reset snd.nxt in retransmit and on triplicate ack.
+
+	if(hdr.ctl & ACK && (seqdiff(hdr.ack, c->snd.last) > 0 || seqdiff(hdr.ack, c->snd.una) < 0)) {
+		debug("Packet ack seqno out of range: hdr.ack=%u snd.una=%u snd.nxt=%u snd.last=%u\n",
+			hdr.ack, c->snd.una, c->snd.nxt, c->snd.last);
+		// Ignore unacceptable RST packets.
+		if(hdr.ctl & RST)
+			return 0;
+		goto reset;
+	}
+
+	c->snd.wnd = hdr.wnd; // TODO: move below
+
+	// 2. Advance snd.una and update retransmit timer
+	// process acks even when hdr.seq doesn't match to adapt early and
+	// get triplicate ack check work even when on both ends packets are not acceptable
 
-	bool acceptable;
+	uint32_t prevrcvnxt = c->rcv.nxt;
+	uint32_t advanced = 0;
+
+	if(hdr.ctl & ACK)
+	{
+		advanced = seqdiff(hdr.ack, c->snd.una);
+
+		if(advanced) {
+			// RTT measurement
+			if(c->rtt_start.tv_sec) {
+				if(c->rtt_seq == hdr.ack) {
+					struct timeval now, diff;
+					gettimeofday(&now, NULL);
+					timersub(&now, &c->rtt_start, &diff);
+					update_rtt(c, diff.tv_sec * 1000000 + diff.tv_usec);
```

Comment:
I was thinking more in "oh, let's triple those times for some obscure reason". Also writing 1000000 in code often makes you vulnerable for miscounting 0s. Like once you'll enter 100000 and will have a very long and annoying debugging session.